### PR TITLE
Create a new Metanorma Document

### PR DESCRIPTION
**Pull request by abunashir**

### CODE REVIEW

**Comment by ronaldtse** [comment]
Few changes needed (feel free to merge this first) following #36.

1. 
```sh
exe/metanorma-new new -t csd -d standard my-csd-document
```

Now creates this structure:
```
├── my-csd-document
│   ├── Gemfile
│   ├── Makefile
│   ├── Makefile.win
│   ├── appveyor.yml
│   ├── deploy.sh
│   └── standard
│       ├── README.adoc
│       ├── cc-document.adoc
│       └── sections
│           ├── 01-scope.adoc
│           ├── 02-norm-refs.adoc
│           ├── 03-termsdef.adoc
│           ├── 04-content.adoc
│           ├── 0a-foreword.adoc
│           ├── 0b-intro.adoc
│           ├── a1-annex-1.adoc
│           └── zz-references.adoc
```

I'd like this instead:
```
├── my-csd-document
│   ├── Gemfile
│   ├── Makefile
│   ├── Makefile.win
│   ├── README.adoc
│   ├── appveyor.yml
│   ├── cc-document.adoc
│   ├── deploy.sh
│   └── sections
│       ├── 01-scope.adoc
│       ├── 02-norm-refs.adoc
│       ├── 03-termsdef.adoc
│       ├── 04-content.adoc
│       ├── 0a-foreword.adoc
│       ├── 0b-intro.adoc
│       ├── a1-annex-1.adoc
│       └── zz-references.adoc
```

2.
In addition, some people like a "flattened" document, i.e. the `include`s should be flattened. Or, we can provide another set of templates that are "flattened".

We will want to use a command line argument for this because it is now very clear that these two audiences ("hierarchical!", "single file!") are very different.

e.g.
```
├── my-csd-document
│   ├── Gemfile
│   ├── Makefile
│   ├── Makefile.win
│   ├── README.adoc
│   ├── appveyor.yml
│   ├── cc-document.adoc <<< All separate sections are subsumed into here
│   ├── deploy.sh
```

3. 

We probably want a command line argument to allow user to pull from a their own custom "template repository". e.g.

```
exe/metanorma-new new -t foo -d bar --template-repository=https://github.com/foo/metanorma-foo-templates my-special-document
```

4. 
In generator.rb:
Screen Shot 2019-03-29 at 12 13 13 PM

Maybe we don't even need to hard code this? e.g. Homebrew's `brew tap [username]/[reponame] assumes the formula (git) repo is already at `https://github.com/[username]/homebrew-[reponame]`.

**Comment by ronaldtse** [comment]
@abunashir please also add the OGC templates: https://github.com/metanorma/mn-templates-ogc (`flavor = :ogc`).

Part of me thinks that the definition of `templates_sources` should be located in each of the gems themselves and loaded through a Template Registry. Thoughts?

**Comment by ronaldtse** [comment]
We're going to change the command line to:
```sh
metanorma new -t {type} -d {template-name} [my-new-document-name]
# or
metanorma new {type}/{template-name} [my-new-document-name]
```

Which will create a directory at `{my-new-document-name}`.

**Comment by abunashir** [comment]
Hi @ronaldtse, there are couple of concerns, let's me try to add those in your comments, and then we can take it from there.

> Few changes needed (feel free to merge this first) following #36.
> 
> 
> ```shell
> exe/metanorma-new new -t csd -d standard my-csd-document
> ```
> 
> Now creates this structure:
> 
> ```
> ├── my-csd-document
> │   ├── Gemfile
> │   ├── Makefile
> │   ├── Makefile.win
> │   ├── appveyor.yml
> │   ├── deploy.sh
> │   └── standard
> │       ├── README.adoc
> │       ├── cc-document.adoc
> │       └── sections
> │           ├── 01-scope.adoc
> │           ├── 02-norm-refs.adoc
> │           ├── 03-termsdef.adoc
> │           ├── 04-content.adoc
> │           ├── 0a-foreword.adoc
> │           ├── 0b-intro.adoc
> │           ├── a1-annex-1.adoc
> │           └── zz-references.adoc
> ```
> I'd like this instead:
> 
> ```
> ├── my-csd-document
> │   ├── Gemfile
> │   ├── Makefile
> │   ├── Makefile.win
> │   ├── README.adoc
> │   ├── appveyor.yml
> │   ├── cc-document.adoc
> │   ├── deploy.sh
> │   └── sections
> │       ├── 01-scope.adoc
> │       ├── 02-norm-refs.adoc
> │       ├── 03-termsdef.adoc
> │       ├── 04-content.adoc
> │       ├── 0a-foreword.adoc
> │       ├── 0b-intro.adoc
> │       ├── a1-annex-1.adoc
> │       └── zz-references.adoc
> ```


*This looks good to me for single type of document (eg: standard/collection), is there any chance a user might want to add multiple type of document in one project, eg: standard and admin combined, then what be structure for this flatten hiaracy?*




> 
> In addition, some people like a "flattened" document, i.e. the `include`s should be flattened. Or, we can provide another set of templates that are "flattened".
> 
> We will want to use a command line argument for this because it is now very clear that these two audiences ("hierarchical!", "single file!") are very different.
> 
> e.g.
> 
> ```
> ├── my-csd-document
> │   ├── Gemfile
> │   ├── Makefile
> │   ├── Makefile.win
> │   ├── README.adoc
> │   ├── appveyor.yml
> │   ├── cc-document.adoc <<< All separate sections are subsumed into here
> │   ├── deploy.sh
> ```

*Yap, I agree on this one, we can add a command line argument later and thats how we can allow both type of user the flexibility to use different structure.*
 
> 
> We probably want a command line argument to allow user to pull from a their own custom "template repository". e.g.
> 
> ```
> exe/metanorma-new new -t foo -d bar --template-repository=https://github.com/foo/metanorma-foo-templates my-special-document
> ```
> 
> In generator.rb:
> Screen Shot 2019-03-29 at 12 13 13 PM
> 
> Maybe we don't even need to hard code this? e.g. Homebrew's `brew tap [username]/[reponame] assumes the formula (git) repo is already at `[https://github.com/[username]/homebrew-[reponame]`](https://github.com/%5Busername%5D/homebrew-%5Breponame%5D%60).

*Yap, that's my whole idea to allow user to provide templates as an argument, but for our tempaltes I think we still should have one based hard coded sources as a fall back, so in any case it would work  and we also make sure the structrue and everythig is right otherwise there would be lot more reason that could beak our things.*

Please have a look and let me know your thoughts on those, thanks!

**Comment by abunashir** [comment]
> @abunashir please also add the OGC templates: https://github.com/metanorma/mn-templates-ogc (`flavor = :ogc`).

I assume the above once is done, right?

> 
> Part of me thinks that the definition of `templates_sources` should be located in each of the gems themselves and loaded through a Template Registry. Thoughts?

Since, it's only a link to template repos, so I feel like it might be much more straight forward to have the base list of templates directly here, otherwise we just add additional depencey to another repo, don't you think?

**Comment by opoudjis** [comment]
Btw, I'm doing Metanorma gem releases now, but given this is WIP, I'm not releasing metanorma-cli

**Comment by ronaldtse** [comment]
> This looks good to me for single type of document (eg: standard/collection), is there any chance a user might want to add multiple type of document in one project, eg: standard and admin combined, then what be structure for this flatten hiaracy?

No, a user will only want a single template for now. It will be very rare for people to want different components from different document types, but if they really wanted to, they can download the templates and manually do so :wink: (we should add this in the documentation).

> Yap, that's my whole idea to allow user to provide templates as an argument, but for our tempaltes I think we still should have one based hard coded sources as a fall back, so in any case it would work and we also make sure the structrue and everythig is right otherwise there would be lot more reason that could beak our things.

Agree for now.

> I assume the above once is done, right? ("OGC")

Yes!

Thanks!

**Comment by ronaldtse** [comment]
@abunashir also would it be better to use https://github.com/ruby-git/ruby-git rather than calling the `git` executable directly?

**Comment by abunashir** [comment]
> In addition, some people like a "flattened" document, i.e. the `include`s should be flattened. Or, we can provide another set of templates that are "flattened".
> 
> We will want to use a command line argument for this because it is now very clear that these two audiences ("hierarchical!", "single file!") are very different.
> 


@ronaldtse: My thought in this one, if we can have some reliable lirabary that could flatten the include directive without messing up the format or anything that would be perfect. 

Otherwise I would say maybe it might be safe have another template file like `csd-flatten.adoc` in the template directory and we use that one when user pass the flatten option, What do you think? Are you aware of any such libararay?


cc: @opoudjis, @ribose-jeffreylau

**Comment by ronaldtse** [comment]
> @ronaldtse: My thought in this one, if we can have some reliable lirabary that could flatten the include directive without messing up the format or anything that would be perfect.

I think what @opoudjis thinks is that we can do it. 

> Otherwise I would say maybe it might be safe have another template file like csd-flatten.adoc in the template directory and we use that one when user pass the flatten option, What do you think? 

However it might be easiest to have a separate "flattened" template. Or, when we publish the templates repository, we automatically generate the flattened template.

How should we name them?

* Hierarchical: `{template}`
* Flattened: `{template}-flat}`

?

**Comment by ronaldtse** [comment]
@abunashir is this ready to merge?

**Comment by ronaldtse** [comment]
@abunashir the specs are now properly done, but the failures are real functionality gaps / bugs. Could you please help fix them? Thanks!

**Comment by abunashir** [comment]
Hi @ronaldtse, fixing the specs and also migrating to Aruba was casing some issue I believe, so I've extracted those here: https://github.com/metanorma/metanorma-cli/pull/41, and once this one is done then let's revisit that again and we can do the migration.

**Comment by ronaldtse** [comment]
@abunashir the issues are not caused by Aruba but are real issues with the CLI. Can you please help fix them?

```
  1) Metanorma processes metanorma options inside Asciidoc creates proper output
     Failure/Error: expect(exist?("test.html")).to be true
       expected true
            got false

  2) Metanorma file with blank header processes an asciidoc ISO document creates proper output
     Got 3 failures from failure aggregation block.
...

  3) Metanorma file with blank header processes all extensions of an asciidoc ISO document creates proper output
...

  4) Metanorma file with configured header extracts isodoc options from Asciidoc file creates proper output
...
```

These are all caused by the CLI's handling of the `-x {ext,ext...}` option.


```
  5) Metanorma file with configured header wraps HTML output wraps with directories
     Got 3 failures from failure aggregation block.
     # ./spec/metanorma_spec.rb:127:in `block (4 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:47:in `block (2 levels) in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:45:in `block in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/aruba-0.14.9/lib/aruba/rspec.rb:22:in `block (2 levels) in <top (required)>'
     5.1) Failure/Error: expect(last_command_started).to be_successfully_executed
            expected "metanorma -w -t iso test.adoc" to be successfully executed
          # ./spec/metanorma_spec.rb:128:in `block (5 levels) in <top (required)>'
     5.2) Failure/Error: expect(exist?("test/test.html")).to be true
            expected true
                 got false
          # ./spec/metanorma_spec.rb:129:in `block (5 levels) in <top (required)>'
     5.3) Failure/Error: expect(exist?("test.alt/test.alt.html")).to be true
            expected true
                 got false
          # ./spec/metanorma_spec.rb:130:in `block (5 levels) in <top (required)>'
```

This is caused by the `-w` "wrapping" option not working.

```
  8) Metanorma file with configured header exports bibdata generates Relaton file
     Failure/Error: expect(last_command_started).to be_successfully_executed
       expected "metanorma -R testrelaton.xml -t iso test.adoc" to be successfully executed
     # ./spec/metanorma_spec.rb:183:in `block (5 levels) in <top (required)>'
     # ./spec/metanorma_spec.rb:182:in `block (4 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:47:in `block (2 levels) in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:45:in `block in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/aruba-0.14.9/lib/aruba/rspec.rb:22:in `block (2 levels) in <top (required)>'
```

This is caused by the `-R {file}` option not working.

```
  9) Metanorma file with configured header exports assets extracts assets in specified folder
     Got 2 failures from failure aggregation block.
     # ./spec/metanorma_spec.rb:228:in `block (4 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:47:in `block (2 levels) in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:45:in `block in <module:RSpecCommand>'
     # ./vendor/bundle/ruby/2.5.0/gems/aruba-0.14.9/lib/aruba/rspec.rb:22:in `block (2 levels) in <top (required)>'
     9.1) Failure/Error: expect(exist?("extract/sourcecode/sourcecode-0000.txt")).to be true
            expected true
                 got false
          # ./spec/metanorma_spec.rb:234:in `block (5 levels) in <top (required)>'
     9.2) Failure/Error: expect("extract/sourcecode/sourcecode-0000.txt").to have_file_content EXPECTED_CONTENT
            expected "extract/sourcecode/sourcecode-0000.txt" to have file content: "def ruby(x)\n  if x < 0 && x > 1\n    return\n  end\nend\n"
          # ./spec/metanorma_spec.rb:235:in `block (5 levels) in <top (required)>'
```

This is caused by the `extract` option not working.

```
  10) Metanorma warns when no file provided should have output: /Need to specify a file to process/
      Failure/Error: it { expect(last_command_started).to have_output(/Need to specify a file to process/) }
        expected "ERROR: \"metanorma compile\" was called with no arguments\nUsage: \"metanorma compile FILENAME\"" to have output: /Need to specify a file to process/
        Diff:
        @@ -1,2 +1,3 @@
        -/Need to specify a file to process/
        +ERROR: "metanorma compile" was called with no arguments
        +Usage: "metanorma compile FILENAME"
      # ./spec/metanorma_spec.rb:297:in `block (3 levels) in <top (required)>'
      # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:47:in `block (2 levels) in <module:RSpecCommand>'
      # ./vendor/bundle/ruby/2.5.0/gems/rspec-command-1.0.3/lib/rspec_command.rb:45:in `block in <module:RSpecCommand>'
      # ./vendor/bundle/ruby/2.5.0/gems/aruba-0.14.9/lib/aruba/rspec.rb:22:in `block (2 levels) in <top (required)>'
```

This is caused by a difference in error message expectation (new CLI uses default Thor message).

**Comment by abunashir** [comment]
Yap, there are quite some issues on the CLI arguments/interface, I'm working on fixing those first

**Comment by abunashir** [comment]
Hi @ronaldtse, this latest changes should fix all of the existing functionality, could you please give it a try when the test passes?

**Comment by ronaldtse** [comment]
Thanks @abunashir ! There's a single spec still failing on AppVeyor:

```
  1) Metanorma processes metanorma options inside Asciidoc
530     Failure/Error: expect(File.exist?("test.xml")).to be true
531
532       expected true
533            got false
534     # ./spec/metanorma_spec.rb:8:in `block (2 levels) in <top (required)>'
535
536Finished in 32.59 seconds (files took 0.69932 seconds to load)
53720 examples, 1 failure
538
539Failed examples:
540
541
```

Everything else is working -- it's strange that the "first spec" has this basic failure 😉 Could you help take a look at this? Thanks!

**Comment by abunashir** [comment]
Hi @ronaldtse, all the issue should be fixed now, could you please double check and merge it when you are happy with these changes :) Thanks!

**Comment by opoudjis** [comment]
This has just missed the window for gem releases, but I can do a release out of sequence if you let me know; I'm already going to have to do one for ietfbib.

File: .travis.yml [hunk 1]
```diff
@@ -13,6 +13,8 @@ before_install:
   - gem install bundler -v 2.0.1
   - bundle update
   - unset _JAVA_OPTIONS
+  - mkdir ./tmp
```
Comment (ronaldtse) [inline .travis.yml]: @abunashir can we not create the `./tmp` here but use `$TMPDIR` given by the system?

File: spec/acceptance/new_spec.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+require "spec_helper"
+
+RSpec.describe "Metanorma" do
+  describe "new document" do
+    it "creates a new metanorma document" do
+      allow(Metanorma::Cli::Generator).to receive(:run)
+
+      command = %w(new -t iso -d standard ./tmp/my-iso-doc)
```
Comment (ronaldtse) [inline spec/acceptance/new_spec.rb]: Here we should use something like:
```ruby
require 'tmpdir'
Dir.mktmpdir do |dir|
  puts "My new temp dir: #{dir}"
end
```

File: .travis.yml (hunk 1 above)
Comment (ronaldtse) [reply .travis.yml]: (see below comments)

File: spec/metanorma/cli/generator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+require "spec_helper"
+
+RSpec.describe Metanorma::Cli::Generator do
+  describe ".run" do
+    context "without existing templates" do
+      it "downloads and generate new document" do
+        document = "./tmp/my-document"
```
Comment (ronaldtse) [inline spec/metanorma/cli/generator_spec.rb]: As above on `Dir.mktmpdir`.

File: spec/metanorma_spec.rb [hunk 4]
```diff
@@ -148,10 +148,12 @@ def ruby(x)
   its(:stdout) { is_expected.to include "bogus_format format is not supported for this standard" }
 end
 
-RSpec.describe "warns when no file provided" do
-  command "metanorma -t iso -x html"
-  its(:stdout) { is_expected.to include "Need to specify a file to process" }
-end
+# Maybe use the thor default error in this case?"
+#
+# RSpec.describe "warns when no file provided" do
```
Comment (ronaldtse) [inline spec/metanorma_spec.rb]: Can we customize Thor's error message? We can use the Thor default for now.

**Comment by ronaldtse** [comment]
I'm going to first merge this and make a new issue for the requested changes, since they are only relevant to testing. Thanks @abunashir !

File: .travis.yml (hunk 1 above)
Comment (abunashir) [reply .travis.yml]: @ronaldtse: What are the benefits of using a shared system's temp directory over project-specific `tmp` directory?

We are only using this directory for the test purpose, and to me, it ensures we can only write to this tmp directly and no other program's tmp file can possibly cause a failing test here and we also ignore this the git, so what are disadvantage we are considering here?

File: spec/metanorma_spec.rb (hunk 4 above)
Comment (abunashir) [reply spec/metanorma_spec.rb]: I will look into it, thanks :)

File: .travis.yml (hunk 1 above)
Comment (ronaldtse) [reply .travis.yml]: The disadvantage is that we now have a `.travis.yml` that specifically relates to these two specs that use this directory. I'd much rather to have that directory done inside the spec. (We want the travis files to be identical across all gems).

Comment (abunashir) [reply .travis.yml]: Okay, what about we remove it from the travis configuration and add RSpec configure block that will only run once before the tests and create the temp directory (if none exists)?

```ruby
config.before(:suite) do
  # create non-existance tmp directory
end
```
